cunumeric/module.py

```diff
@@ -8137,6 +8137,430 @@ def percentile(
     )
 
 
+# args:
+#
+# arr:      [in] source nd-array on which quantiles are calculated;
+#                NaNs ignored; precondition: assumed sorted!
+# q_arr:    [in] quantile input values nd-array;
+# axis:     [in] axis along which quantiles are calculated;
+# method:   [in] func(q, n) returning (gamma, j),
+#                where = array1D.size;
+# keepdims: [in] boolean flag specifying whether collapsed axis
+#                should be kept as dim=1;
+# to_dtype: [in] dtype to convert the result to;
+# qs_all:   [in/out] result pass through or created (returned)
+#
+def nanquantile_impl(
+    arr: np.ndarray,
+    q_arr: np.ndarray[Any],
+    axis: Optional[int],
+    axes_set: Iterable[int],
+    original_shape: tuple[int, ...],
+    method: Callable[[float, int], tuple[float, int]],
+    keepdims: bool,
+    to_dtype: np.dtype[Any],
+    qs_all: Optional[np.ndarray],
+) -> np.ndarray:
+    ndims = len(arr.shape)
+
+    if axis is None:
+        if keepdims:
+            remaining_shape = (1,) * len(original_shape)
+        else:
+            remaining_shape = ()  # only `q_arr` dictates shape;
+        # quantile applied to `arr` seen as 1D;
+    else:
+        # arr.shape -{axis}; if keepdims use 1 for arr.shape[axis]:
+        # (can be empty [])
+        #
+        if keepdims:
+            remaining_shape = tuple(
+                1 if k in axes_set else original_shape[k]
+                for k in range(0, len(original_shape))
+            )
+        else:
+            remaining_shape = tuple(
+                arr.shape[k] for k in range(0, ndims) if k != axis
+            )
+
+    # compose qarr.shape with arr.shape:
+    #
+    # result.shape = (q_arr.shape, arr.shape -{axis}):
+    #
+    qresult_shape = (*q_arr.shape, *remaining_shape)
+
+    # construct result Np.Ndarray, non-flattening approach:
+    #
+    if qs_all is None:
+        qs_all = zeros(qresult_shape, dtype=to_dtype)
+    else:
+        # implicit conversion from to_dtype to qs_all.dtype assumed
+        #
+        if qs_all.shape != qresult_shape:
+            raise ValueError("wrong shape on output array")
+
+    # ndarray of non-NaNs:
+    #
+    non_nan_counts = count_nonzero((isnan(arr) == False), axis=axis)
+    assert non_nan_counts.shape == remaining_shape
+
+    arr_ones = ones(remaining_shape, dtype=arr.dtype)
+
+    arr_gammas = zeros(remaining_shape, dtype=arr.dtype)
+    arr_lvals = zeros(remaining_shape, dtype=arr.dtype)
+    arr_rvals = zeros(remaining_shape, dtype=arr.dtype)
+
+    for qindex, q in np.ndenumerate(q_arr):
+        assert qs_all[qindex].shape == remaining_shape
+
+        for aindex, n in np.ndenumerate(non_nan_counts):
```

Comment:
IIUC this is iterating over all the elements of `non_nan_counts` one at a time, which is not great. It would be better if we could "vectorize" this operation, i.e. operate over all the `non_nan_counts` concurrently (which should be feasible, since the original computation is strictly element-wise).

I think the original `quantile_impl` is actually pretty close to vectorizable already (the operations it does on a single `n` can be lifted to full-array operations). Something like this patch may be sufficient to allow `quantile_impl` to handle both the `ignore_nans=False` case (its original function), and `ignore_nans=True` (i.e. the equivalent to `nanquantile_impl`):

```
diff --git a/cunumeric/module.py b/cunumeric/module.py
index fbeadf48..ce40a39d 100644
--- a/cunumeric/module.py
+++ b/cunumeric/module.py
@@ -7804,9 +7804,14 @@ def quantile_impl(
         if qs_all.shape != qresult_shape:
             raise ValueError("wrong shape on output array")

+    if ignore_nan:
+        non_nan_counts = count_nonzero(logical_not(isnan(arr)), axis=axis)
+        assert non_nan_counts.shape == remaining_shape
+    else:
+        non_nan_counts = n
+
     for index, q in np.ndenumerate(q_arr):
-        (gamma, j) = method(q, n)
-        (left_pos, right_pos) = (j, j + 1)
+        (gamma, pos) = method(q, non_nan_counts)

         # (N-1) dimensional ndarray of left, right
         # neighbor values:
@@ -7814,17 +7819,13 @@ def quantile_impl(
         # non-flattening approach:
         #
         # extract values at index=left_pos;
-        arr_1D_lvals = arr.take(left_pos, axis)
-        arr_vals_shape = arr_1D_lvals.shape
+        arr_1D_lvals = arr.take(pos, axis)

-        if right_pos >= n:
-            # some quantile methods may result in j==(n-1),
-            # hence (j+1) could surpass array boundary;
-            #
-            arr_1D_rvals = zeros(arr_vals_shape, dtype=arr_1D_lvals.dtype)
-        else:
-            # extract values at index=right_pos;
-            arr_1D_rvals = arr.take(right_pos, axis)
+        # extract values at index=right_pos;
+        # some quantile methods may result in j==(n-1),
+        # hence (j+1) could surpass array boundary;
+        pos[pos < non_nan_counts] += 1
+        arr_1D_rvals = arr.take(pos, axis)

         # vectorized for axis != None;
         # (non-flattening approach)
```

Of course, we'd also need to change all the quantile methods to be able to operate elementwise on arrays, rather than single elements only. For example, this would be the implementation of `inverted_cdf`:

```
def inverted_cdf(q: float, n: int | ndarray) -> tuple[ndarray, ndarray]:
    pos = q * n
    k = floor(pos)
    g = pos - k
    gamma = where(g > 0, 1.0, 0.0)
    j = k.astype(np.int64) - 1
    return (
        where(j < 0, 0.0, gamma),
        where(j < 0, 0, j),
    )
```

Reply:
I think I understand better this suggestion. 

I'm a bit hesitant to change the `quantile()` implementation, which has been deployed ("Open-Close Principle: keep code opened to addition but closed to modification"...as much as possible). Specifically, I would prefer to not modify the interpolation primitives for `quantile()` (if that actually was the suggestion).

But I could add separate interpolation primitives for the "nan-tiles" case, along the lines suggested above.

Reply:
Per latest conversation w/ @manopapad , this vectorization may require a more advanced slicing that might not even be available in `numpy` yet. Likely to be addressed in a separate future PR.

Reply:
Left a comment as future work